src/Stage/ReadStage.php

```diff
@@ -0,0 +1,95 @@
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Stage;
+
+use ApiPlatform\Core\DataProvider\CollectionDataProviderInterface;
+use ApiPlatform\Core\DataProvider\ItemDataProviderInterface;
+use ApiPlatform\Core\DataProvider\OperationDataProviderTrait;
+use ApiPlatform\Core\DataProvider\SubresourceDataProviderInterface;
+use ApiPlatform\Core\Exception\InvalidIdentifierException;
+use ApiPlatform\Core\Exception\NotFoundException;
+use ApiPlatform\Core\Exception\RuntimeException;
+use ApiPlatform\Core\Identifier\IdentifierConverterInterface;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\ToggleableOperationAttributeTrait;
+
+/**
+ * Retrieves data from the applicable data provider and sets it as a request parameter called data.
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+final class ReadStage implements ReadStageInterface
+{
+    use OperationDataProviderTrait;
+    use ToggleableOperationAttributeTrait;
+
+    public function __construct(CollectionDataProviderInterface $collectionDataProvider, ItemDataProviderInterface $itemDataProvider, SubresourceDataProviderInterface $subresourceDataProvider = null, IdentifierConverterInterface $identifierConverter = null, ResourceMetadataFactoryInterface $resourceMetadataFactory = null)
+    {
+        $this->collectionDataProvider = $collectionDataProvider;
+        $this->itemDataProvider = $itemDataProvider;
+        $this->subresourceDataProvider = $subresourceDataProvider;
+        $this->identifierConverter = $identifierConverter;
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(array $attributes, array $parameters, ?array $filters, string $method, array $normalizationContext)
+    {
+        if (!isset($attributes['resource_class'])
+            || !$attributes['receive']
+            || ('POST' === $method && isset($attributes['collection_operation_name']))
+            || $this->isOperationAttributeDisabled($attributes, self::OPERATION_ATTRIBUTE_KEY)
+        ) {
+            return null;
+        }
+
+        $context = null === $filters ? [] : ['filters' => $filters];
+        $context += $normalizationContext ?? [];
+
+        if (isset($attributes['collection_operation_name'])) {
+            return $this->getCollectionData($attributes, $context);
+        }
+
+        $data = [];
+
+        if ($this->identifierConverter) {
+            $context[IdentifierConverterInterface::HAS_IDENTIFIER_CONVERTER] = true;
+        }
+
+        try {
+            $identifiers = $this->extractIdentifiers($parameters, $attributes);
+
+            if (isset($attributes['item_operation_name'])) {
+                $data = $this->getItemData($identifiers, $attributes, $context);
+            } elseif (isset($attributes['subresource_operation_name'])) {
+                // Legacy
```

Comment:
Why legacy? It's a valid use case to not inject a `SubResourceDataPprovider` (especially if you use the library directly, without the Symfony bridge).

Reply:
Not my code here but I can remove the comment. The `SubresourceData` is mandatory in the GraphQL read stage.